core/trino-main/src/main/java/io/trino/operator/TopNProcessor.java

```diff
@@ -76,30 +87,46 @@ public void addInput(Page page)
 
     public Page getOutput()
     {
-        if (outputIterator == null) {
+        if (outputIterator == null && topNBuilder != null) {
             // start flushing
             outputIterator = topNBuilder.buildResult();
         }
 
-        Page output = null;
-        if (outputIterator.hasNext()) {
+        Page output;
+        if (outputIterator != null && outputIterator.hasNext()) {
             output = outputIterator.next();
         }
         else {
-            outputIterator = emptyIterator();
+            closeTopNBuilder();
+            return null;
         }
         updateMemoryReservation();
         return output;
     }
 
     public boolean noMoreOutput()
     {
-        return outputIterator != null && !outputIterator.hasNext();
+        return outputIterator == null;
     }
 
     private void updateMemoryReservation()
     {
-        requireNonNull(topNBuilder, "topNBuilder is null");
-        localUserMemoryContext.setBytes(topNBuilder.getEstimatedSizeInBytes());
+        if (topNBuilder == null) {
+            localMemoryContext.setBytes(0);
+            return;
+        }
+        localMemoryContext.setBytes(topNBuilder.getEstimatedSizeInBytes());
```

Comment:
Assuming you add the getEstimatedSizeInBytes() suggestion below, you can just do:
```java
localMemoryContext.setBytes(getEstimatedSizeInBytes())
```

Reply:
Done